proof/encoding.go

```diff
@@ -478,19 +480,40 @@ func GroupKeyRevealEncoder(w io.Writer, val any, _ *[8]byte) error {
 }
 
 func GroupKeyRevealDecoder(r io.Reader, val any, buf *[8]byte, l uint64) error {
+	// Return early if the val is not a pointer to a GroupKeyReveal.
```

Comment:
So this is now actually a _pointer_ to an interface?

Reply:
If `val` is asserted to be of type `GroupKeyReveal` rather than `*GroupKeyReveal`, and `val == nil`, the type assertion inside `GroupKeyRevealDecoder` (e.g., `typ, ok := val.(GroupKeyReveal)`) will fail. This happens because Go's type assertion mechanism requires both a dynamic type and a value to confirm the type, and a `nil` value with no dynamic type results in `ok == false`.

For example:

```go
var gkrDecoded GroupKeyReveal
var scratchBuffDecode [8]byte
err := GroupKeyRevealDecoder(
    &buffer, gkrDecoded, &scratchBuffDecode,
    uint64(buffer.Len()),
)
```

This will fail because `GroupKeyRevealDecoder` cannot determine whether `val` implements the `GroupKeyReveal` interface when `val == nil`. Since `gkrDecoded` is not a pointer, it has no dynamic type in this context, making the assertion invalid.

By contrast, if `val` is of type `*GroupKeyReveal`, the dynamic type is always present, even when the value is `nil`. This allows `GroupKeyRevealDecoder` to qualify `val == nil` correctly.

I think there might be a way around this using the `reflect` package but I haven't gotten it to work yet.